mono/utils/mono-dl.c

```diff
@@ -143,6 +148,26 @@ get_dl_name_from_libtool (const char *libtool_file)
 	return line;
 }
 
+#ifdef ENABLE_NETCORE
+static const char *
+fix_libc_name (const char *name)
+{
+	if (strncmp (name, "libc", 4) == 0) {
+		// Taken from CoreCLR: https://github.com/dotnet/coreclr/blob/6b0dab793260d36e35d66c82678c63046828d01b/src/pal/src/loader/module.cpp#L568-L576
+#if defined (HOST_DARWIN)
+		return "/usr/lib/libc.dylib";
```

Comment:
A bit surprised this one hardcodes an absolute path.

Will this work on iOS?

Reply:
**TL;DR**: It should be fine on iOS/tvOS/watchOS too.

On macOS:
```console
$ ls -l /usr/lib/libc.dylib
lrwxr-xr-x  1 root  wheel  15 Oct  1  2018 /usr/lib/libc.dylib -> libSystem.dylib
$ ls -l /usr/lib/libSystem.dylib
lrwxr-xr-x  1 root  wheel  17 Oct  1  2018 /usr/lib/libSystem.dylib -> libSystem.B.dylib
$ ls -l /usr/lib/libSystem.B.dylib
-rwxr-xr-x  1 root  wheel  60880 Aug 23 08:30 /usr/lib/libSystem.B.dylib
```
So the `libc.dylib` symlink ends up in `libSystem.B.dylib` which is suspiciously small to be honest.  All the symbols you would expect from `libc` seem to be in `/usr/lib/dyld`, which is the dynamic linker itself:
```console
$ file /usr/lib/dyld
/usr/lib/dyld: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamic linker x86_64] [i386:Mach-O dynamic linker i386]
/usr/lib/dyld (for architecture x86_64):        Mach-O 64-bit dynamic linker x86_64
/usr/lib/dyld (for architecture i386):  Mach-O dynamic linker i386
```
Let's add some verbose logging:
```console
$ DYLD_PRINT_LIBRARIES=y ls
dyld: /bin/ls
dyld: /usr/lib/libutil.dylib
dyld: /usr/lib/libncurses.5.4.dylib
dyld: /usr/lib/libSystem.B.dylib
[...]
```
No trace of `libc.dylib`.

So what's the situation on iOS? Not 100% sure, as I don't have direct shell access, there are however libs part of Xcode to support debugging:
```console
$ pwd
/Users/lewurm/Library/Developer/Xcode/iOS DeviceSupport/13.1.2 (17A860)/Symbols/usr/lib
$ ls -lha libc.dylib
ls: libc.dylib: No such file or directory
$ ls -lha libSystem.B.dylib
-rw-r--r--  1 lewurm  staff    12K Oct  2 10:13 libSystem.B.dylib
$ file dyld
dyld: Mach-O 64-bit dynamic linker arm64
```
My guess is that the iOS device will contain the symlinks as we have seen above on macOS. Let's snoop around with `lldb` on an iOS device:
```
(lldb) settings set target.env-vars DYLD_PRINT_LIBRARIES=y
(lldb) b xamarin_main
Breakpoint 1: where = sdbdebugtest`::xamarin_main(int, char **, XamarinLaunchMode) + 72 at monotouch-main.m:230:28, address = 0x0000000100412700
(lldb) r
Process 364 launched: '/Users/lewurm/Projects/sdbdebugtest/sdbdebugtest/bin/iPhone/Debug/device-builds/iphone9.3-13.1.2/sdbdebugtest.app/sdbdebugtest' (arm64)
dyld: <7D0702F8-6706-341E-AC5E-C51EA5735484> /private/var/containers/Bundle/Application/6B85FA1F-3D0F-4382-8EC2-4BC85661D704/sdbdebugtest.app/sdbdebugtest
dyld: <9B025821-A606-3634-A038-3032B850F8E2> /System/Library/Frameworks/UIKit.framework/UIKit
 <0DF2911E-80CB-3289-8A1E-ED0913D55A12> /System/Library/Frameworks/Foundation.framework/Foundation
dyld: <DD159500-7D53-3484-AA3D-9DD5073D7A2F> /System/Library/Frameworks/CFNetwork.framework/CFNetwork
Library/Frameworks/GSS.framework/GSS
dyld: <F8C60E32-C706-3653-AD1F-6BE0E1A67550> /usr/lib/libz.1.dylib
dyld: <C7159C1D-C5F2-3938-8019-5F1F26A812D6> /usr/lib/libiconv.2.dylib
dyld: <F11CB1CC-B616-39B1-8D76-192163AF9C98> /usr/lib/libSystem.B.dylib
[...]
```
Evidently it loads `libSystem.B.dylib` as well. Let's use `dlopen` directly:
```
(lldb) call (void *) dlopen("/usr/lib/libc.dylib", 0)
(void *) $0 = 0x000000000f490e00
(lldb) call (void *) dlopen("/usr/lib/libSystem.B.dylib", 0)
(void *) $2 = 0x000000000f490e00
(lldb) call (void *) dlopen("/usr/lib/libSystem.dylib", 0)
(void *) $3 = 0x000000000f490e00
(lldb) call (void *) dlopen("/usr/lib/libASDF.dylib", 0)
(void *) $4 = 0x0000000000000000
```
They all return the same address. Looks good to me!